### ISSUE TITLE
The multiple recipients case is underdefined

### ISSUE THREAD

**Issue by AnomalRoil** [NeilMadden/jose-ecdh-1pu#4]
In [Appendix B.](https://datatracker.ietf.org/doc/html/draft-madden-jose-ecdh-1pu-04#appendix-B), the draft 4 is currently saying the following (emphasis mine):

> Alice is sending an identical message to Bob and Charlie.  Because Bob and Charlie are using the same curve (X25519), Alice reuses the same ephemeral key-pair for both recipients and includes it in the JWE Protected Header.  **If this was not the case, Alice _should_ generate a separate ephemeral key-pair for each recipient and include it in each _per-recipient header instead_.**

However, the current specification isn't properly covering the case where we have multiple recipients using different kind of keys on different curves, because in that case the _static_ sender key also needs to match the recipient's _static_ keys in a pairwise manner and also the ephemeral key. This is needed to determine the proper Ze and Zs value used to derive the key decrypting the "encrypted_key" used in key wrapping mode.

The current spec also doesn't cover the notion of per-recipient headers properly: the "epk" is usually a "protected header" field, whereas the current notion of "per-recipient header" seems to be **unprotected headers** only.

It is not clear from the language above found in appendix B. how the multiple-recipient case with different key type should be handled, where would which data go, what are per-recipients headers, etc.

This could be solved by using per recipient "skid" values, for example, and introducing a notion of per-recipient protected headers, but the whole per-recipients story should probably be part of the spec explicitly, since it has already caused [lengthy discussions](https://github.com/decentralized-identity/didcomm-messaging/issues/218#issuecomment-889996191).

What is your position on that topic @NeilMadden ?

PS: Notice the "skid" header is also not fully specified in the sense that it is not clear whether it should be a protected or an unprotected header and it is not used in any of the examples in the appendixes. (But this is probably to be discussed rather in #3.)

**Comment by NeilMadden** [comment]
It's a good point. I think really it is probably best if we constrain all recipients to use keys that are on the same curve as the sender. Trying to do anything else would require multiple different sender static keys, and it quickly becomes very complex (and likely error-prone). This would also ensure the `epk` can always go in the protected header section.

That does make `ECDH-1PU` special compared to existing JWE encryption algorithms though, where you can currently happily mix and match different algorithms for different recipients. But I think that is probably an acceptable limitation as the other encryption algorithms currently defined have very different security characteristics.

**Comment by AnomalRoil** [comment]
> Trying to do anything else would require multiple different sender static keys

Yes, indeed.

> it quickly becomes very complex (and likely error-prone)

Well, from an implementer POV, the logic to support the different key types is already there.
So it's just a matter of having per-recipient `skid` and `epk` fields or using more complex data struct for these fields if we are keeping them in the general headers.

What it would need is just a proper description, I'd say. The underlying complexity is already there.

**Comment by NeilMadden** [comment]
OK, I think I agree (sadly). I'll try and come up with some wording, but I'm on vacation for a little while now. PRs always welcome.

**Comment by AnomalRoil** [comment]
I can try and come up with something.

Shall we use a [JWK Set](https://datatracker.ietf.org/doc/html/rfc7517#section-5) in the `epk` field and simply mandate each recipient to select the key in that set matching their `kid`'s key type?
I believe the `skid` field doesn't necessarily needs to be protected as long as the `apu` field is protected, so I'd be fine adding a per-recipient `skid` field in the per-recipient unprotected headers.

What do you think?

The alternative would be to introduce a "protected per-recipient header" and have the per-recipient `skid` there.

In general I think we don't want a per-recipient `epk` because if we are sending a message to 10 people 4 of which use Curve25519 keys, 3 of which P-256 and the rest P-384, then we _only_ need 3 `epk` and the `epk` type is fully described by the `crv` field in the JWK that are part of a JWK set. So no need to duplicate the `epk` in each per-recipient header and the recipient knows which `epk` to use based on the `crv` used by their specified `kid` in their own header.

**Comment by NeilMadden** [comment]
I don't think you could change the `epk` field to a JWK Set now, as it's a standard field defined to contain just one JWK. I also don't think we can introduce a per-recipient protected header section easily, but I'm uneasy about putting the `epk` in an unprotected field. But I guess that is what you have to do for `ECDH-ES` at the moment? (I've never had a use-case of sending a JWE to multiple recipients with different algorithms).

Including the ephemeral public key in the Concat KDF input data should be sufficient to protect the `epk` from tampering, and is really required to prevent malleability anyway, so I may adjust the `ECDH-1PU` spec to mandate its inclusion (along with the static PKs while I'm at it). In that case it would be fine to include the `epk` in the unprotected per-recipient header sections. As you say, it's less efficient but I don't know how much we really need to optimise this case. Hopefully most sane applications will pick a single curve and stick to that.

**Comment by AnomalRoil** [comment]
Yeah, I'm re-reading [4.6.1.1 of RFC 7518](https://datatracker.ietf.org/doc/html/rfc7518#section-4.6.1.1) and I think you're right that it sadly mandates a single JWK key in the `epk` field:
> 4.6.1.1.  "epk" (Ephemeral Public Key) Header Parameter
> 
>    The "epk" (ephemeral public key) value created by the originator for
>    the use in key agreement algorithms.  This key is represented as a
>    JSON Web Key [JWK] public key value.  It MUST contain only public key
>    parameters and SHOULD contain only the minimum JWK parameters
>    necessary to represent the key; other JWK parameters included can be
>    checked for consistency and honored, or they can be ignored.  This
>    Header Parameter MUST be present and MUST be understood and processed
>    by implementations when these algorithms are used.

Now regarding the inclusion of the per-recipient `epk` in the KDF, I'm not convinced it's necessary to prevent tampering, but I would also prefer to have the `epk` we are using being in the AAD for peace of mind... See below for a wild idea how to work around it.
 
> Including the ephemeral public key in the Concat KDF input data should be sufficient to protect the epk from tampering

I guess we could say that the `epk` is already "somewhat" included in the `Ze` component that is being fed to the Concat KDF...
Since we're also already including the auth tag produced using the CEK in the KDF (in the multiple-recipient case since we're in key-wrapping mode then).

I think it would require some careful thinking about what kind of threat or attack we are trying to mitigate by including the `epk` in the per-recipient KEK derivation, but I'm not convinced it's actually necessary.

Notice that an attacker able to change an unprotected `epk` field could also change the auth tag field itself... or any other unprotected data easily, including the encrypted keys themselves. What really "saves" us here is the secret static key of party U allowing to "authenticate" the KEX for party V, which is the Zs part of the Concat KDF input, which means we're already solely relying on the Concat KDF derivation to authenticate the key agreement. I don't see why we'd need to add an extra "integrity check" for the `epk`.

## Wild idea

Now, a way to "hack" the existing spec to include multiple `epk` in the AAD might be the `apu` field: it looks like a good place to store a base64url encoded JSON structure including a JWK Set... Any implementation of ECDH-1PU will already have to "handle" the `apu` base64url encoded field using special logic, since we are "recommending" to have data in there... so we might take advantage of that and have more logic relying on that field.

The RFC7518 isn't saying this isn't right as far as I understand:
>    The "apu" (agreement PartyUInfo) value for key agreement algorithms
>    using it (such as "ECDH-ES"), represented as a base64url-encoded
>    string.  When used, the PartyUInfo value contains information about
>    the producer.  Use of this Header Parameter is OPTIONAL.  This Header
>    Parameter MUST be understood and processed by implementations when
>    these algorithms are used.

And it would allow to avoid having duplicated `epk` fields for all recipients using the same curve.
(Typically if we have 10 people using X25519 and 6 persons using P-256, this would save us 14*256 bits of duplicated unprotected headers.)

## Another necessary change 

Finally, one thing we might want to really mandate is **public key validation of the recipients static public keys**. Because when we are in a multi-recipient setting where we **re-use** the same `epk` for multiple recipients, we definitively want to make sure our recipients' static keys are all valid to prevent invalid curve attacks that would compromise the whole message because of one rogue recipient.

This is currently "hinted" in Section 4, but it just says:

> care should be taken to ensure that the public key is a valid point on the same curve as the private key

without actually mandating it using proper _requirements terminology_. But that's really a MUST, IMO.

Key validation is especially important whenever we speak of `epk` re-use to encrypt _different messages_ for different parties, usually, **which is not our case** since we are encrypting the same CEK for everyone... But still, "small-subgroup style" invalid curve attacks where you have multiple recipient with invalid public keys might allow to learn bits of both our ephemeral and our static secret keys I think... (And it might be worse if ECDH-1PU were to be used in an interactive/automated way with some kind of oracle in case of errors.) 
So we really want to have key validation enforced I think.

**Comment by NeilMadden** [comment]
> Now regarding the inclusion of the per-recipient epk in the KDF, I'm not convinced it's necessary to prevent tampering

Generally speaking, ECDH-based encryption is malleable without this step. For example, for some elliptic curves there are multiple equivalent public keys that will compute the same shared secret (e.g., by adding points of small order to a legitimate public key). This is only what cryptographers refer to as "benign malleability", but it's best not to ignore it. Sadly, the existing `ECDH-ES` algorithm doesn't mandate including the `epk` in the KDF input. I may change that for `ECDH-1PU`.

> I guess we could say that the epk is already "somewhat" included in the Ze component that is being fed to the Concat KDF... Since we're also already including the auth tag produced using the CEK in the KDF (in the multiple-recipient case since we're in key-wrapping mode then).

That's a fair point, and I think that would indeed be sufficient to ensure non-malleability for the multi-recipient case - but only if the `epk` is included in the Protected Header, which gets us back to that being shared between all recipients...

I'd like to also fix it for the single recipient (no key-wrapping) case, but then `ECDH-ES` technically has the same issue so the inconsistency may not be worth it.

> I think it would require some careful thinking about what kind of threat or attack we are trying to mitigate by including the epk in the per-recipient KEK derivation, but I'm not convinced it's actually necessary.

IMO the way modern cryptographic constructions should work is that you pick your security goals and then design constructions that provably(*) meet those goals, and only relax if you need to (e.g. for efficiency). For `ECDH-1PU` the security goal is authenticated encryption, which implies non-malleability. Not including the `epk` in the KDF calculation makes the JWT malleable, which means it doesn't achieve its stated security goal. Now, this might only be a "benign" violation of the security goal, but I'd rather avoid such a step if I can.

(*) Not that I've actually proved `ECDH-1PU` correct as yet...

Re invalid curve attacks - I agree. I'll strengthen the discussion in the Security Considerations to add a MUST wording. (Although, IIRC, `X25519`/`X448` require no validation as all possible inputs are valid).

**Comment by NeilMadden** [comment]
(I'll think more about the `apu` suggestion - but I guess we'd at least need a way to avoid colliding with other uses of that header)
